src/ecs/resources/input.rs

```diff
@@ -62,45 +92,73 @@ impl<'a> Iterator for PressedMouseButtons<'a> {
 /// ```
 #[derive(Default)]
 pub struct InputHandler {
-    pressed_keys: HashMap<VirtualKeyCode, KeyQueryState>,
-    pressed_mouse_buttons: HashMap<MouseButton, KeyQueryState>,
+    pressed_keys: Vec<VirtualKeyCode>,
+    previous_pressed_keys: Vec<VirtualKeyCode>,
+    pressed_mouse_buttons: Vec<MouseButton>,
+    previous_pressed_mouse_buttons: Vec<MouseButton>,
     // defined to match glium
     mouse_position: Option<(i32, i32)>,
     previous_mouse_position: Option<(i32, i32)>,
+    axes: HashMap<i32, Axis>,
+    actions: HashMap<i32, Button>,
+    text_this_frame: String,
 }
 
 impl InputHandler {
     /// Creates a new input handler.
     pub fn new() -> InputHandler {
         InputHandler {
-            pressed_keys: HashMap::default(),
-            pressed_mouse_buttons: HashMap::default(),
+            pressed_keys: Vec::new(),
```

Comment:
Abstracting `pressed_*` and `previous_*` into a separate struct could reduce code duplication.